File: 1917-maximum-average-pass-ratio/maximum-average-pass-ratio.py
class Solution:
    def maxAverageRatio(self, classes: List[List[int]], extraStudents: int) -> float:
        
        heap=[]
        for passs,total in classes:
            answer=((passs+1)/(total+1))-(passs/total)
            heapq.heappush(heap,(-answer,passs,total))
        for _ in range(extraStudents):
            res,pstud,total=heapq.heappop(heap)
            pstud+=1
            total+=1
            answer=((pstud+1)/(total+1))-(pstud/total)
            heapq.heappush(heap,
            (-answer,pstud,total))
        
        result=0
        return sum((passs/total) for calc,passs,total in heap)/len(classes)


         

File: 1917-maximum-average-pass-ratio/maximum-average-pass-ratio.java
import java.util.*;

class Solution {
    
    public double calc(double p, double t) {
        return p / t;
    }


    public double gain(double p, double t) {
        return calc(p + 1, t + 1) - calc(p, t);
    }

    public double maxAverageRatio(int[][] classes, int extraStudents) {
       
        PriorityQueue<double[]> heap = new PriorityQueue<>(
            (a, b) -> Double.compare(b[0], a[0]) // sort by gain descending
        );

     
        for (int[] c : classes) {
            double pass = c[0];
            double total = c[1];
            heap.offer(new double[]{gain(pass, total), pass, total});
        }

      
        for (int i = 0; i < extraStudents; i++) {
            double[] top = heap.poll();  
            double pass = top[1] + 1;
            double total = top[2] + 1;
            heap.offer(new double[]{gain(pass, total), pass, total});
        }

      
        double sum = 0.0;
        while (!heap.isEmpty()) {
            double[] cur = heap.poll();
            sum += calc(cur[1], cur[2]);
        }

        return sum / classes.length;
    }
}
